python/ray/serve/_private/http_proxy.py

```diff
@@ -535,19 +449,191 @@ async def __call__(self, scope, receive, send):
+                # some replicas crash simultaneously (e.g. if the head node crashes),
+                # requests to the dead replicas hang until the gcs recovers.
+                # This asyncio.wait can kill those hanging requests and retry them
+                # at another replica. Release tests should kill the head node and
+                # check if latency drops significantly. See
+                # https://github.com/ray-project/ray/pull/29534 for more info.
+                _, request_timed_out = await asyncio.wait(
+                    [object_ref], timeout=RAY_SERVE_REQUEST_PROCESSING_TIMEOUT_S
+                )
+                if request_timed_out:
+                    logger.info(
+                        "Request didn't finish within "
+                        f"{RAY_SERVE_REQUEST_PROCESSING_TIMEOUT_S} seconds. Retrying "
+                        "with another replica. You can modify this timeout by "
+                        'setting the "RAY_SERVE_REQUEST_PROCESSING_TIMEOUT_S" env var.'
+                    )
+                    backoff = True
+                else:
+                    result = await object_ref
+                    break
+            except asyncio.CancelledError:
+                # Here because the client disconnected, we will return a custom
+                # error code for metric tracking.
+                return DISCONNECT_ERROR_CODE
+            except RayTaskError as e:
+                error_message = f"Unexpected error, traceback: {e}."
+                await Response(error_message, status_code=500).send(
+                    scope, receive, send
+                )
+                return "500"
+            except RayActorError:
+                logger.info(
+                    "Request failed due to replica failure. There are "
+                    f"{HTTP_REQUEST_MAX_RETRIES - retries} retries "
+                    "remaining."
+                )
+                backoff = True
+            if backoff:
+                await asyncio.sleep(backoff_time_s)
+                # Be careful about the expotential backoff scaling here.
+                # Assuming 10 retries, 1.5x scaling means the last retry is 38x the
+                # initial backoff time, while 2x scaling means 512x the initial.
+                backoff_time_s *= 1.5
+                retries += 1
+                backoff = False
+        else:
+            error_message = f"Task failed with {HTTP_REQUEST_MAX_RETRIES} retries."
+            await Response(error_message, status_code=500).send(scope, receive, send)
+            return "500"
+
+        if isinstance(result, (starlette.responses.Response, RawASGIResponse)):
+            await result(scope, receive, send)
+            return str(result.status_code)
+        else:
+            await Response(result).send(scope, receive, send)
+            return "200"
+
+    async def proxy_asgi_receive(self, receive: Receive, queue: asyncio.Queue):
+        """TODO"""
```

Comment:
Could you elaborate or remove the `TODO`?

Reply:
done